### ISSUE TITLE
Flagrum throws exception on startup due to certain Steam Workshop mods

### ISSUE THREAD

**Issue by jtigano** [Kizari/Flagrum#82]
After installing the program I get the following error whenever I try to start it:

A Tasks exception(s) were not observed either by waiting on the task or accessing its exception properly.  As a result, the unobserved exception was rethrown by the finalizer thread. (Exception of type 'Flagrum.Web.Utilities.UserFriendlyException' was thrown.)

The application will now close.

**Comment by Kizari** [comment]
This is usually caused by an old faulty Steam Workshop mod being present on the system. First thing would be to unsubscribe to any really old ones, especially those from SE themselves. You can try turn each one off one at a time until Flagrum will launch to determine which causes the issue, or you could just turn them all off and try enable them one at a time until it breaks.

If you have any local Steam Workshop mods, it may also pay to try remove those if the above doesn't work.

**Comment by Kizari** [comment]
Related to #45

**Comment by Kizari** [comment]
While there can be multiple causes for this (I believe one of them is that the binmods are an older file format), one relatively common one that has popped up multiple times seems to be that some binmod names can have commas in them, which breaks the CSV format (i.e. binmod.list). It seems commas are allowed in names, so SE's parser must have used the quotes surrounding column values to denote boundaries in which commas are not separators. Updating the parser in Flagrum to behave in a similar manner would fix this issue for those particular binmods.

Some examples of binmods that are affected to help with testing:
```
Online_Accessory_Sandogasa_(Male,accessory).ffxvbinmod
Online_Accessory_Muffler_and_hat_(Male,accessory).ffxvbinmod
Online_Accessory_Arachne_(Female,Face_accessory).ffxvbinmod
```

Further to this, perhaps it isn't worth crashing Flagrum over these issues? It seems to cause pretty big problems to non-tech-savvy users, so perhaps a little extra effort to warn of the broken mods that can't be displayed would be sufficient, and it could otherwise function as normal just without those mods in the mod list. This should only apply to the very few binmods (I think only the official ones?) that use the old binmod format that isn't worth the time reversing just for a couple of mods to display in Flagrum that barely anyone uses. Even catching these individual errors and displaying appropriate warnings would incur quite a bit of effort though, so it would be worth weighing that up when deciding to go ahead on the solution.

Will aim to address this in Flagrum 1.5, but as that is an absolutely massive update, it may be delayed until a future patch.

**Comment by Kizari** [comment]
https://github.com/JoshClose/CsvHelper seems to parse this file fine, so long as a header row is prepended to the CSV. While doing this to the file directly is likely not a good idea as it may break it for the game, it can be done in memory before parsing the file—which with a tiny text file like this it isn't an issue to load the whole thing in memory.

Here is an example of doing this with the above library:
```csharp
const string path = @"C:\Users\Kizari\Documents\My Games\FINAL FANTASY XV\Steam\MySteamID\mod\binmod.list";

// Read the whole file into memory
var lines = File.ReadAllLines(path).ToList();

// Insert the header row as the first line
lines.Insert(0, "Path,IsActive,Fixid,IsRemote");

// Convert the strings back into bytes so they can be used in a memory stream
var final = string.Join('\n', lines);
var bytes = Encoding.UTF8.GetBytes(final);

// Parse the CSV
using var stream = new MemoryStream(bytes);
using var reader = new StreamReader(stream);
using var csv = new CsvReader(reader, CultureInfo.InvariantCulture);
var records = csv.GetRecords<Binmod>().ToList();
```

The above sample assumes there is also a class with the following declaration in the codebase:
```csharp
public class Binmod
{
    public string? Path { get; set; }
    public bool IsActive { get; set; }
    public uint Fixid { get; set; }
    public bool IsRemote { get; set; }
}
```

**Comment by Doodlacie** [comment]
> This is usually caused by an old faulty Steam Workshop mod being present on the system. First thing would be to unsubscribe to any really old ones, especially those from SE themselves. You can try turn each one off one at a time until Flagrum will launch to determine which causes the issue, or you could just turn them all off and try enable them one at a time until it breaks.
> 
> If you have any local Steam Workshop mods, it may also pay to try remove those if the above doesn't work.

I have unsubscribed from all workshop mods and it still shows "Flagrum.Web.Utilities.UserFriendlyException" was thrown

**Comment by Kizari** [comment]
Do you have any local binmods on your machine? They would be located in:
C:\\Users\\**YOUR_NAME**\\Documents\\My Games\\FINAL FANTASY XV\\Steam\\**YOUR_STEAM_ID**\\mod
(Replace the bold parts with the correct values. Also if your operating system isn't on C:, use whatever drive it is on).

If so, delete them or remove them from this folder and then try again.

If not, please attach your full crash log so I can look into it. You can find it at:
```
%LOCALAPPDATA%\Flagrum\crash.txt
```

**Comment by jtigano** [comment]
I removed all steam mods, deleted and reinstalled the game, and only installed flagrum and it seemed to solve the problem.  The issue was related to an old mod.

**Comment by CallItDivineRetribution** [comment]
I unsubscribed from all Steam mods, uninstalled and reinstalled FFXV. Uninstalled and reinstalled Flagrum. I slowly resubscribed to some Steam mods, it was working fine at first but I guess I dumbly subscribed to a bad one because the error is back. I did the same steps again and haven't resubscribed to any Steam mods yet, but it's still showing the error this time.

The location of my game is in
C:\Program Files (x86)\Steam\steamapps\common\FINAL FANTASY XV
not in documents and there is no mods folder

[crash.txt](https://github.com/Kizari/Flagrum/files/12425010/crash.txt)

**Comment by Kizari** [comment]
@CallItDivineRetribution Your crash log shows there's a bad mod still listed in your binmod.list file. You can find the location of that in Flagrum's settings tab. You can delete that file and then launch FFXV to generate a new one that doesn't contain any of the mods you unsubscribed from. Once that is regenerated, Flagrum should work fine. If you're still getting this error, it's either because you still have a bad mod in the same folder with binmod.list, or because Steam has not unsubscribed you from one of the bad mods properly.

The issue with most of these older Steam Workshop mods will be fixed in Flagrum 1.5, but that's still a couple of weeks away or so.

**Comment by CallItDivineRetribution** [comment]
Thank you! I think I got it now. I opened FFXV to make sure it showed that I didn't still have any mods active, then tried opening Flagrum again afterwards and it didn't throw the error this time.

**Comment by Kizari** [comment]
This has been fixed in Flagrum 1.5.
